### PR TITLE
Stop asking headers if no new headers are presented

### PR DESCRIPTION
**Pull request by dangershony**
When a node is in IBD and the timer is triggered and will starts a resync or a new block was produced and relayed to our node (which can not be connected and will also trigger a Resync) to prevent a continuance loop of headers that are already asked we stop asking more headers if the received headers could not connect.

Here is the main flow:
- node in IBD triggered a getheaders and the locator points to consensus tip
- peer responds with 2000 headers
- node received and connects the new headers 
- node send again a getheaders and the locator points to the last received header
- peer responds with the next 2000 headers
-this continues til the tip when the peer will respond with zero headers.

*timer* 
- timer start and triggeres a resync
- a new getheaders is sent to the peer
- continue on the main flow
- and a second loop starts in parallel with the other loop

*new block*
- a new block is produced on the network 
- a single header is sent to out peer
- the header is too far ahead we cant connect and we triggere a resync
- an additional parallel loop of asking for headers starts and does not stop until the tip

### CODE REVIEW

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 1]
```diff
@@ -367,6 +367,15 @@ protected virtual async Task ProcessHeadersAsync(INetworkPeer peer, List<BlockHe
...
                     this.cachedHeaders.AddRange(headers);
                     this.logger.LogDebug("All {0} items were not consumed and added to cache.", headers.Count);
+                    this.logger.LogTrace("(-)[HEADERS_NOT_CONSUMED]");
+
+                    return;
+                }
+
+                if (result.Consumed == this.BestReceivedTip)
+                {
+                    this.logger.LogDebug("No new headers where presented");
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: `were`

Comment (codingupastorm) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: Not to steal your thunder but I'm all about this new feature atm

```suggestion
                    this.logger.LogDebug("No new headers were presented");
```

**Comment by rowandh** [comment]
@dangershony this might need to be rebased against latest master so that the integration tests pass deterministically 😄

**Comment by noescape00** [comment]
This is not needed since when we are trying to sync we will send block locator which is based on the best header received from the peer. So each time we are asking for headers we are not aware about and we provide info about what we are aware about. 

So the change doesn't make sense for the peers that follow the protocol. 

For the malicious peers it doesn't make sense because if malicious peer would like to spam with useless headers he will just have to send any other header but not the best we know about.

**Comment by dangershony** [comment]
>this violates btc protocol that states that node should stop asking for headers when an empty headers payload is received

How this violates the protocol exactly?
Instead of stopping at a zero headers response the node stops asking headers when no new headers are presented.

**Comment by dangershony** [comment]
>This is not needed since when we are trying to sync we will send block locator which is based on the best header received from the peer.

Yeah but in this case we just ask for a the same locator again based on the same best header that was received, its a completely useless request that just generates more traffic.

**Comment by dangershony** [comment]
Closing for now as its not urgent, we will revisit this later.

**Pull request closed without merging**